File: clients/ios/Sources/MotoViewKit/MotoViewCore.swift
//
//  MotoViewCore.swift — the typed Swift facade over the Rust C ABI.
//
//  Every method here calls a `mv_ffi_*` C function (declared in motoview_ffi.h,
//  implemented in client/src/ffi.rs), decodes the `{ok,value}`/`{ok,error}` JSON
//  envelope, and surfaces a Swift `Result`/throwing API. The Rust core does ALL
//  the real work — parsing, the keyed LIS diff, the chain-key cert verification.
//  Swift never reimplements any of it; this is the dumb-hands / smart-brain split
//  the web client already uses, applied to native.
//

import Foundation
import MotoViewFFI

public enum MotoViewError: Error, Equatable {
    /// The Rust core returned `{"ok":false,"error":"..."}`.
    case core(String)
    /// The envelope itself was not the shape we expected (should never happen).
    case malformed(String)
}

/// The MotoView native core. Stateless; every call is a pure function over the
/// Rust FFI. Hold one instance (or use the static helpers) per app.
public struct MotoViewCore {

    public init() {}

    // MARK: - Envelope plumbing

    /// Call a 1-string FFI function, free the returned C string, and return the
    /// decoded envelope's `value` (as Foundation JSON Any) or throw on `ok:false`.
    private func call1(_ input: String,
                       _ fn: (UnsafePointer<CChar>?) -> UnsafeMutablePointer<CChar>?) throws -> Any {
        let raw: UnsafeMutablePointer<CChar>? = input.withCString { cstr in
            fn(cstr)
        }
        return try decodeEnvelope(raw)
    }

    /// Decode an FFI envelope C string into its `value`, freeing the buffer.
    private func decodeEnvelope(_ raw: UnsafeMutablePointer<CChar>?) throws -> Any {
        guard let raw = raw else { throw MotoViewError.malformed("null FFI result") }
        defer { mv_ffi_string_free(raw) }
        let json = String(cString: raw)
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw MotoViewError.malformed("non-JSON envelope: \(json)")
        }
        if let ok = obj["ok"] as? Bool, ok == true {
            guard let value = obj["value"] else {
                throw MotoViewError.malformed("envelope missing 'value'")
            }
            return value
        }
        let reason = (obj["error"] as? String) ?? "unknown"
        throw MotoViewError.core(reason)
    }

    // MARK: - IR parsing

    /// Parse a JSON IR forest (the Ir.mo.toJsonForest wire format) into a UINode
    /// tree, running the REAL Rust parser (depth guard, escape decoding, schema).
    public func parseForest(_ json: String) throws -> [UINode] {
        let value = try call1(json) { mv_ffi_parse_forest($0) }
        guard let arr = value as? [Any] else {
            throw MotoViewError.malformed("forest value not an array")
        }
        return try arr.map { try UINode.fromJSONObject($0) }
    }

    /// Parse a single JSON IR node into a UINode.
    public func parseNode(_ json: String) throws -> UINode {
        let value = try call1(json) { mv_ffi_parse_node($0) }
        return try UINode.fromJSONObject(value)
    }

    /// Render a JSON IR forest to HTML (used by the `.raw` fallback / debugging).
    public func renderForest(_ json: String) throws -> String {
        let value = try call1(json) { mv_ffi_render_forest($0) }
        guard let s = value as? String else {
            throw MotoViewError.malformed("render value not a string")
        }
        return s
    }

    // MARK: - Keyed diff

    /// Diff two JSON IR forests with the same keyed-reconcile/LIS the web brain
    /// uses, returning a `DiffPlan` (full re-render or an ordered op list).
    public func irDiff(old oldJSON: String, new newJSON: String) throws -> DiffPlan {
        let raw: UnsafeMutablePointer<CChar>? = oldJSON.withCString { o in
            newJSON.withCString { n in
                mv_ffi_ir_diff(o, n)
            }
        }
        let value = try decodeEnvelope(raw)
        return try DiffPlan.fromJSONObject(value)
    }

    // MARK: - Chain-key certificate verification

    /// Verify an IC response against the PINNED NNS root key (mainnet). Returns
    /// the certified `/time` (ns) as a `UInt64`-safe `String` (it is a u128 in
    /// Rust). Throws `.core("<CertError>")` on any failure (fails closed).
    ///
    /// `path` is the certified-data tree path, e.g.
    /// `[Data("canister".utf8), canisterId, Data("certified_data".utf8)]`.
    @discardableResult
    public func verifyResponse(cert: Data,
                               canister: Data,
                               path: [Data],
                               body: Data,
                               nowNs: UInt128Pair,
                               maxOffsetNs: UInt64 = 0) throws -> String {
        // Encode the path as length-prefixed segments: [u32_le len][bytes]...
        var pathBuf = Data()
        for seg in path {
            var len = UInt32(seg.count).littleEndian
            withUnsafeBytes(of: &len) { pathBuf.append(contentsOf: $0) }
            pathBuf.append(seg)
        }

        let raw: UnsafeMutablePointer<CChar>? = cert.withUnsafeBytes { certPtr in
            canister.withUnsafeBytes { canPtr in
                pathBuf.withUnsafeBytes { pathPtr in
                    body.withUnsafeBytes { bodyPtr in
                        mv_ffi_verify_response(
                            certPtr.bindMemory(to: UInt8.self).baseAddress, cert.count,
                            canPtr.bindMemory(to: UInt8.self).baseAddress, canister.count,
                            pathPtr.bindMemory(to: UInt8.self).baseAddress, pathBuf.count,
                            bodyPtr.bindMemory(to: UInt8.self).baseAddress, body.count,
                            nowNs.hi, nowNs.lo, maxOffsetNs
                        )
                    }
                }
            }
        }
        let value = try decodeEnvelope(raw)
        guard let obj = value as? [String: Any], let timeNs = obj["time_ns"] as? String else {
            throw MotoViewError.malformed("verify value missing 'time_ns'")
        }
        return timeNs
    }
}

/// A 128-bit nanosecond timestamp split into two u64 halves, matching the FFI
/// (C has no u128). Build from a Foundation `Date` via `.now()`.
public struct UInt128Pair {
    public let hi: UInt64
    public let lo: UInt64
    public init(hi: UInt64, lo: UInt64) {
        self.hi = hi
        self.lo = lo
    }
    /// Nanoseconds since the Unix epoch for `date`, as a hi/lo pair. The value
    /// fits in 64 bits until the year ~2554, so `hi` is 0 in practice — but we
    /// keep the split so the ABI is future-proof and matches the u128 Rust side.
    public static func nanos(since date: Date = Date()) -> UInt128Pair {
        let ns = UInt64(date.timeIntervalSince1970 * 1_000_000_000)
        return UInt128Pair(hi: 0, lo: ns)
    }
}

File: clients/android/motoview/src/main/kotlin/dev/motoview/core/MotoViewCore.kt
package dev.motoview.core

import org.json.JSONArray
import org.json.JSONObject

/*
 * MotoViewCore.kt — the Kotlin facade over the Rust core, the Android sibling of
 * the iOS MotoViewCore.swift.
 *
 * SCAFFOLD ONLY — NOT BUILT ON THIS MACHINE. There is no Android NDK /
 * ANDROID_HOME / cargo-ndk here, so the `.so` is not produced and this module is
 * not compiled. See clients/android/CARGO_NDK.md for the exact build steps.
 *
 * Binding strategy: the Rust crate exposes a flat C ABI (mv_ffi_*), not JNI-named
 * symbols. Two supported options, both documented in CARGO_NDK.md:
 *   (A) UniFFI — add the uniffi proc-macro to client/, `cargo run --bin
 *       uniffi-bindgen generate ... --language kotlin`, and the generated Kotlin
 *       replaces these `external` declarations with a typed binding. (Preferred
 *       once uniffi-bindgen is on PATH.)
 *   (B) A tiny JNI shim (jni_bridge.c, ~30 lines) that re-exports each mv_ffi_*
 *       as a `Java_dev_motoview_core_MotoViewNative_*` symbol, built into the
 *       same `libmotoview_client.so` by cargo-ndk. The `external fun`s below
 *       match that shim's names.
 *
 * Either way the surface is identical to iOS: parse a JSON IR forest into a
 * UINode tree, diff two forests, render to HTML, and verify a response against
 * the pinned NNS root key — all decided in Rust.
 */

/** JNI entry points implemented by the Rust core's JNI shim (see CARGO_NDK.md). */
internal object MotoViewNative {
    init {
        // Loads libmotoview_client.so from src/main/jniLibs/<abi>/ (placed there
        // by cargo-ndk). Throws UnsatisfiedLinkError if the .so is absent — which
        // it is on this machine, by design (scaffold not built here).
        System.loadLibrary("motoview_client")
    }

    external fun parseForest(json: String): String
    external fun parseNode(json: String): String
    external fun renderForest(json: String): String
    external fun irDiff(oldJson: String, newJson: String): String
    external fun verifyResponse(
        cert: ByteArray,
        canister: ByteArray,
        path: ByteArray,
        body: ByteArray,
        nowNsHi: Long,
        nowNsLo: Long,
        maxOffsetNs: Long
    ): String
}

class MotoViewError(message: String) : Exception(message)

/** Stateless facade — every call is a pure function over the Rust FFI. */
object MotoViewCore {

    private fun envelopeValue(raw: String): Any {
        val obj = JSONObject(raw)
        if (obj.optBoolean("ok", false)) {
            return obj.get("value")
        }
        throw MotoViewError(obj.optString("error", "unknown"))
    }

    fun parseForest(json: String): List<UINode> {
        val value = envelopeValue(MotoViewNative.parseForest(json)) as JSONArray
        return (0 until value.length()).map { UINode.fromJson(value.getJSONObject(it)) }
    }

    fun parseNode(json: String): UINode {
        val value = envelopeValue(MotoViewNative.parseNode(json)) as JSONObject
        return UINode.fromJson(value)
    }

    fun renderForest(json: String): String =
        envelopeValue(MotoViewNative.renderForest(json)) as String

    fun irDiff(oldJson: String, newJson: String): DiffPlan {
        val value = envelopeValue(MotoViewNative.irDiff(oldJson, newJson)) as JSONObject
        return DiffPlan.fromJson(value)
    }

    /**
     * Verify an IC response against the PINNED NNS root key (mainnet). Returns the
     * certified /time (ns) as a String (u128 in Rust). Throws on failure (fails
     * closed). `path` segments are length-prefixed [u32_le len][bytes]... here.
     */
    fun verifyResponse(
        cert: ByteArray,
        canister: ByteArray,
        path: List<ByteArray>,
        body: ByteArray,
        nowNsHi: Long = 0,
        nowNsLo: Long,
        maxOffsetNs: Long = 0
    ): String {
        val pathBuf = encodePath(path)
        val value = envelopeValue(
            MotoViewNative.verifyResponse(cert, canister, pathBuf, body, nowNsHi, nowNsLo, maxOffsetNs)
        ) as JSONObject
        return value.getString("time_ns")
    }

    private fun encodePath(path: List<ByteArray>): ByteArray {
        val out = ArrayList<Byte>()
        for (seg in path) {
            val n = seg.size
            out.add((n and 0xFF).toByte())
            out.add(((n shr 8) and 0xFF).toByte())
            out.add(((n shr 16) and 0xFF).toByte())
            out.add(((n shr 24) and 0xFF).toByte())
            seg.forEach { out.add(it) }
        }
        return out.toByteArray()
    }
}
